content/blog/2022-05-17-object-oriented-features/index.md

```diff
@@ -0,0 +1,210 @@
+
+## `ThisAugmentor`
+This class keeps track of the current class (since we do not allow nested classes yet, but when we do we will just keep track of stack of classes), and a stack of contexts. When we enter a new context which occurs when we see a new block of code, new function, or enter a new loop block. Any declarations or function arguments add a new binding to the current context. Then isField function can check to see if the binding is in a context. If it is, we know its not a field because the variables in the context would shadow all the fields. But if its not in any of those contexts, then we can lookup the fields in the current class context.
+
+### `FExpr`
+
+If a function expression f() is not already marked as a method call, then we know there is no parent of this node that is a MethodCall. Then if there is a function with the name f in the class it is being called in, then we must convert it to a MethodCall where the object of the MethodCall is this.
+
+### `FunctionDecl`, `FunctionDefn`, `FExpr`
+
+Methods all take the object as a hidden first argument. If the FunctionDecl is a method we add this as the first argument.
+
+We do not need to do this for FunctionDefn since this is encapsulated in the FunctionDecl that belongs to the FunctionDefn.
+
+For all FExprs that are MethodCalls, qualified or unqualified must add this as the first argument.
+
+### `Var`
+
+We must lookup to see if the Var is a field and if it is we wrap it in a FieldAccess
+
+### `SimpleDecl`, `ExtendedDecl`
+
+When we see a decl we must add a new binding in the variable context. This is needed in case there is anything that should shadow the scope of a field. I have not handled this shadowing with function arguments. I wonder if the Var part of the FunctionDecl adds this to the context actually.
+
+# IR Generation
+We introduce three new changes to the IR:
+
+1. Addition of `IRClassDefn`
+2. Add list of `IRClassDefn` to `IRCompUnit`
+3. Addition `IRDataArray`
+
+The `IRClassDefn` class contains its methods in `IRFuncDefn` form and the `IRDataArray` which corresponds to the classes's static dispatch vector. The `IRCompUnit` hoists methods out of each class so assembly can be generated for each method. The intention is that method names all have unique encodings and will never conflict with regular function names. Therefore we can output them as normal functions and all code that is generated will call using the correct encoding. More details on the ABI conventions we came up with to accomplish this below.
```

Comment:
Is there any way to explain what's going on here in a higher-level way? What are the concepts that the IR contains now—not the *classes* it contains now? Without this explanation, the reader is forced to infer the role of `IRDataArray` and stuff from its name.